File: rnz_cropwise_flutter/lib/models/fertilizer_calculation.dart
class NPKValues {
  final double nitrogen;
  final double phosphorus;
  final double potassium;

  NPKValues({
    required this.nitrogen,
    required this.phosphorus,
    required this.potassium,
  });

  factory NPKValues.fromJson(Map<String, dynamic> json) {
    return NPKValues(
      nitrogen: (json['nitrogen'] ?? 0).toDouble(),
      phosphorus: (json['phosphorus'] ?? 0).toDouble(),
      potassium: (json['potassium'] ?? 0).toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'nitrogen': nitrogen,
      'phosphorus': phosphorus,
      'potassium': potassium,
    };
  }

  @override
  String toString() => 'N:$nitrogen, P:$phosphorus, K:$potassium';
}

class ProductRecommendation {
  final String? productId;
  final String productName;
  final String npkRatio;
  final double quantity;
  final String unit;
  final String? applicationMethod;
  final String? timing;
  final double? cost;

  ProductRecommendation({
    this.productId,
    required this.productName,
    required this.npkRatio,
    required this.quantity,
    this.unit = 'kg',
    this.applicationMethod,
    this.timing,
    this.cost,
  });

  factory ProductRecommendation.fromJson(Map<String, dynamic> json) {
    return ProductRecommendation(
      productId: json['productId'],
      productName: json['productName'] ?? '',
      npkRatio: json['npkRatio'] ?? '',
      quantity: (json['quantity'] ?? 0).toDouble(),
      unit: json['unit'] ?? 'kg',
      applicationMethod: json['applicationMethod'],
      timing: json['timing'],
      cost: json['cost']?.toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (productId != null) 'productId': productId,
      'productName': productName,
      'npkRatio': npkRatio,
      'quantity': quantity,
      'unit': unit,
      if (applicationMethod != null) 'applicationMethod': applicationMethod,
      if (timing != null) 'timing': timing,
      if (cost != null) 'cost': cost,
    };
  }
}

class ApplicationStageProduct {
  final String? productId;
  final String productName;
  final double quantity;
  final String unit;

  ApplicationStageProduct({
    this.productId,
    required this.productName,
    required this.quantity,
    this.unit = 'kg',
  });

  factory ApplicationStageProduct.fromJson(Map<String, dynamic> json) {
    return ApplicationStageProduct(
      productId: json['productId'],
      productName: json['productName'] ?? '',
      quantity: (json['quantity'] ?? 0).toDouble(),
      unit: json['unit'] ?? 'kg',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (productId != null) 'productId': productId,
      'productName': productName,
      'quantity': quantity,
      'unit': unit,
    };
  }
}

class ApplicationSchedule {
  final String stage;
  final int daysAfterSowing;
  final List<ApplicationStageProduct> products;
  final String? instructions;

  ApplicationSchedule({
    required this.stage,
    required this.daysAfterSowing,
    required this.products,
    this.instructions,
  });

  factory ApplicationSchedule.fromJson(Map<String, dynamic> json) {
    return ApplicationSchedule(
      stage: json['stage'] ?? '',
      daysAfterSowing: json['daysAfterSowing'] ?? 0,
      products: (json['products'] as List<dynamic>?)
              ?.map((p) => ApplicationStageProduct.fromJson(p))
              .toList() ??
          [],
      instructions: json['instructions'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'stage': stage,
      'daysAfterSowing': daysAfterSowing,
      'products': products.map((p) => p.toJson()).toList(),
      if (instructions != null) 'instructions': instructions,
    };
  }
}

class FertilizerCalculation {
  final String? id;
  final String userId;
  final String? fieldId;
  final String cropType;
  final double fieldSize;
  final String fieldUnit;
  final String soilType;
  final NPKValues currentNPK;
  final double targetYield;
  final String yieldUnit;
  final NPKValues requiredNPK;
  final List<ProductRecommendation> productRecommendations;
  final List<ApplicationSchedule> applicationSchedule;
  final double totalCost;
  final double? costPerAcre;
  final String currency;
  final String? notes;
  final List<String>? recommendations;
  final DateTime calculatedAt;
  final bool savedByUser;
  final bool shared;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  FertilizerCalculation({
    this.id,
    required this.userId,
    this.fieldId,
    required this.cropType,
    required this.fieldSize,
    this.fieldUnit = 'acres',
    required this.soilType,
    required this.currentNPK,
    required this.targetYield,
    this.yieldUnit = 'tons/acre',
    required this.requiredNPK,
    required this.productRecommendations,
    required this.applicationSchedule,
    this.totalCost = 0,
    this.costPerAcre,
    this.currency = 'USD',
    this.notes,
    this.recommendations,
    required this.calculatedAt,
    this.savedByUser = false,
    this.shared = false,
    this.createdAt,
    this.updatedAt,
  });

  factory FertilizerCalculation.fromJson(Map<String, dynamic> json) {
    return FertilizerCalculation(
      id: json['_id'],
      userId: json['userId'] ?? '',
      fieldId: json['fieldId'],
      cropType: json['cropType'] ?? '',
      fieldSize: (json['fieldSize'] ?? 0).toDouble(),
      fieldUnit: json['fieldUnit'] ?? 'acres',
      soilType: json['soilType'] ?? '',
      currentNPK: NPKValues.fromJson(json['currentNPK'] ?? {}),
      targetYield: (json['targetYield'] ?? 0).toDouble(),
      yieldUnit: json['yieldUnit'] ?? 'tons/acre',
      requiredNPK: NPKValues.fromJson(json['requiredNPK'] ?? {}),
      productRecommendations: (json['productRecommendations'] as List<dynamic>?)
              ?.map((p) => ProductRecommendation.fromJson(p))
              .toList() ??
          [],
      applicationSchedule: (json['applicationSchedule'] as List<dynamic>?)
              ?.map((s) => ApplicationSchedule.fromJson(s))
              .toList() ??
          [],
      totalCost: (json['totalCost'] ?? 0).toDouble(),
      costPerAcre: json['costPerAcre']?.toDouble(),
      currency: json['currency'] ?? 'USD',
      notes: json['notes'],
      recommendations: (json['recommendations'] as List<dynamic>?)
          ?.map((r) => r.toString())
          .toList(),
      calculatedAt: DateTime.parse(
        json['calculatedAt'] ?? DateTime.now().toIso8601String(),
      ),
      savedByUser: json['savedByUser'] ?? false,
      shared: json['shared'] ?? false,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : null,
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) '_id': id,
      'userId': userId,
      if (fieldId != null) 'fieldId': fieldId,
      'cropType': cropType,
      'fieldSize': fieldSize,
      'fieldUnit': fieldUnit,
      'soilType': soilType,
      'currentNPK': currentNPK.toJson(),
      'targetYield': targetYield,
      'yieldUnit': yieldUnit,
      'requiredNPK': requiredNPK.toJson(),
      'productRecommendations':
          productRecommendations.map((p) => p.toJson()).toList(),
      'applicationSchedule':
          applicationSchedule.map((s) => s.toJson()).toList(),
      'totalCost': totalCost,
      if (costPerAcre != null) 'costPerAcre': costPerAcre,
      'currency': currency,
      if (notes != null) 'notes': notes,
      if (recommendations != null) 'recommendations': recommendations,
      'calculatedAt': calculatedAt.toIso8601String(),
      'savedByUser': savedByUser,
      'shared': shared,
      if (createdAt != null) 'createdAt': createdAt!.toIso8601String(),
      if (updatedAt != null) 'updatedAt': updatedAt!.toIso8601String(),
    };
  }

  FertilizerCalculation copyWith({
    String? id,
    String? userId,
    String? fieldId,
    String? cropType,
    double? fieldSize,
    String? fieldUnit,
    String? soilType,
    NPKValues? currentNPK,
    double? targetYield,
    String? yieldUnit,
    NPKValues? requiredNPK,
    List<ProductRecommendation>? productRecommendations,
    List<ApplicationSchedule>? applicationSchedule,
    double? totalCost,
    double? costPerAcre,
    String? currency,
    String? notes,
    List<String>? recommendations,
    DateTime? calculatedAt,
    bool? savedByUser,
    bool? shared,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return FertilizerCalculation(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      fieldId: fieldId ?? this.fieldId,
      cropType: cropType ?? this.cropType,
      fieldSize: fieldSize ?? this.fieldSize,
      fieldUnit: fieldUnit ?? this.fieldUnit,
      soilType: soilType ?? this.soilType,
      currentNPK: currentNPK ?? this.currentNPK,
      targetYield: targetYield ?? this.targetYield,
      yieldUnit: yieldUnit ?? this.yieldUnit,
      requiredNPK: requiredNPK ?? this.requiredNPK,
      productRecommendations:
          productRecommendations ?? this.productRecommendations,
      applicationSchedule: applicationSchedule ?? this.applicationSchedule,
      totalCost: totalCost ?? this.totalCost,
      costPerAcre: costPerAcre ?? this.costPerAcre,
      currency: currency ?? this.currency,
      notes: notes ?? this.notes,
      recommendations: recommendations ?? this.recommendations,
      calculatedAt: calculatedAt ?? this.calculatedAt,
      savedByUser: savedByUser ?? this.savedByUser,
      shared: shared ?? this.shared,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

File: src/lib/mongodb/models/FertilizerCalculation.ts
import mongoose, { Schema, Document, Model } from 'mongoose';

// Interface for NPK values
export interface INPKValues {
  nitrogen: number;
  phosphorus: number;
  potassium: number;
}

// Interface for Product Recommendation
export interface IProductRecommendation {
  productId: mongoose.Types.ObjectId;
  productName: string;
  npkRatio: string;
  quantity: number; // in kg
  unit: string;
  applicationMethod: string;
  timing: string;
  cost?: number;
}

// Interface for Application Schedule
export interface IApplicationSchedule {
  stage: string; // e.g., "Basal", "Top Dressing 1", "Top Dressing 2"
  daysAfterSowing: number;
  products: {
    productId: mongoose.Types.ObjectId;
    productName: string;
    quantity: number;
    unit: string;
  }[];
  instructions: string;
}

// Interface for Fertilizer Calculation Document
export interface IFertilizerCalculation extends Document {
  userId: mongoose.Types.ObjectId;
  fieldId?: mongoose.Types.ObjectId;
  
  // Input Parameters
  cropType: string;
  fieldSize: number;
  fieldUnit: 'acres' | 'hectares';
  soilType: string;
  
  // Current NPK levels (from soil test)
  currentNPK: INPKValues;
  
  // Target values
  targetYield: number;
  yieldUnit: string; // e.g., "tons/acre", "quintals/hectare"
  
  // Calculated Requirements
  requiredNPK: INPKValues;
  
  // Recommendations
  productRecommendations: IProductRecommendation[];
  applicationSchedule: IApplicationSchedule[];
  
  // Cost Analysis
  totalCost: number;
  costPerAcre?: number;
  currency: string;
  
  // Additional Info
  notes?: string;
  recommendations?: string[];
  
  // Metadata
  calculatedAt: Date;
  savedByUser: boolean;
  shared: boolean;
  
  createdAt: Date;
  updatedAt: Date;
}

const FertilizerCalculationSchema = new Schema<IFertilizerCalculation>(
  {
    userId: {
      type: Schema.Types.ObjectId,
      ref: 'User',
      required: true,
      index: true,
    },
    fieldId: {
      type: Schema.Types.ObjectId,
      ref: 'Field',
      index: true,
    },
    
    // Input Parameters
    cropType: {
      type: String,
      required: true,
      index: true,
    },
    fieldSize: {
      type: Number,
      required: true,
      min: 0,
    },
    fieldUnit: {
      type: String,
      enum: ['acres', 'hectares'],
      required: true,
      default: 'acres',
    },
    soilType: {
      type: String,
      required: true,
    },
    
    // Current NPK levels
    currentNPK: {
      nitrogen: { type: Number, required: true, default: 0 },
      phosphorus: { type: Number, required: true, default: 0 },
      potassium: { type: Number, required: true, default: 0 },
    },
    
    // Target values
    targetYield: {
      type: Number,
      required: true,
      min: 0,
    },
    yieldUnit: {
      type: String,
      required: true,
      default: 'tons/acre',
    },
    
    // Calculated Requirements
    requiredNPK: {
      nitrogen: { type: Number, required: true },
      phosphorus: { type: Number, required: true },
      potassium: { type: Number, required: true },
    },
    
    // Recommendations
    productRecommendations: [{
      productId: { type: Schema.Types.ObjectId, ref: 'Product' },
      productName: { type: String, required: true },
      npkRatio: { type: String, required: true },
      quantity: { type: Number, required: true },
      unit: { type: String, default: 'kg' },
      applicationMethod: { type: String },
      timing: { type: String },
      cost: { type: Number },
    }],
    
    applicationSchedule: [{
      stage: { type: String, required: true },
      daysAfterSowing: { type: Number, required: true },
      products: [{
        productId: { type: Schema.Types.ObjectId, ref: 'Product' },
        productName: { type: String, required: true },
        quantity: { type: Number, required: true },
        unit: { type: String, default: 'kg' },
      }],
      instructions: { type: String },
    }],
    
    // Cost Analysis
    totalCost: {
      type: Number,
      required: true,
      default: 0,
    },
    costPerAcre: {
      type: Number,
    },
    currency: {
      type: String,
      default: 'USD',
    },
    
    // Additional Info
    notes: String,
    recommendations: [String],
    
    // Metadata
    calculatedAt: {
      type: Date,
      default: Date.now,
    },
    savedByUser: {
      type: Boolean,
      default: false,
    },
    shared: {
      type: Boolean,
      default: false,
    },
  },
  {
    timestamps: true,
  }
);

// Indexes for better query performance
FertilizerCalculationSchema.index({ userId: 1, createdAt: -1 });
FertilizerCalculationSchema.index({ cropType: 1, createdAt: -1 });
FertilizerCalculationSchema.index({ fieldId: 1 });

// Static method to calculate NPK requirements
FertilizerCalculationSchema.statics.calculateNPKRequirements = function(
  cropType: string,
  targetYield: number,
  fieldSize: number,
  currentNPK: INPKValues,
  soilType: string
): INPKValues {
  // NPK requirements per ton of yield (simplified - should be crop-specific)
  const npkPerTon: { [key: string]: INPKValues } = {
    wheat: { nitrogen: 25, phosphorus: 12, potassium: 25 },
    rice: { nitrogen: 20, phosphorus: 10, potassium: 20 },
    corn: { nitrogen: 30, phosphorus: 15, potassium: 30 },
    maize: { nitrogen: 30, phosphorus: 15, potassium: 30 },
    soybean: { nitrogen: 15, phosphorus: 8, potassium: 20 },
    cotton: { nitrogen: 35, phosphorus: 18, potassium: 35 },
    sugarcane: { nitrogen: 40, phosphorus: 20, potassium: 40 },
    potato: { nitrogen: 28, phosphorus: 14, potassium: 35 },
    tomato: { nitrogen: 25, phosphorus: 12, potassium: 30 },
    onion: { nitrogen: 22, phosphorus: 11, potassium: 25 },
    default: { nitrogen: 25, phosphorus: 12, potassium: 25 },
  };

  // Get crop-specific requirements or use default
  const cropRequirements = npkPerTon[cropType.toLowerCase()] || npkPerTon.default;

  // Calculate total required NPK based on target yield
  const totalRequired: INPKValues = {
    nitrogen: cropRequirements.nitrogen * targetYield,
    phosphorus: cropRequirements.phosphorus * targetYield,
    potassium: cropRequirements.potassium * targetYield,
  };

  // Soil type adjustment factor
  const soilAdjustment: { [key: string]: number } = {
    sandy: 1.2,        // More nutrients needed
    loamy: 1.0,        // Optimal
    clay: 0.9,         // Better nutrient retention
    silty: 1.05,       // Moderate
    'black soil': 0.85, // Rich in nutrients
    'red soil': 1.15,   // Poor in nutrients
    default: 1.0,
  };

  const adjustment = soilAdjustment[soilType.toLowerCase()] || soilAdjustment.default;

  // Calculate net required NPK (total required - current available + soil adjustment)
  const netRequired: INPKValues = {
    nitrogen: Math.max(0, (totalRequired.nitrogen - currentNPK.nitrogen) * adjustment),
    phosphorus: Math.max(0, (totalRequired.phosphorus - currentNPK.phosphorus) * adjustment),
    potassium: Math.max(0, (totalRequired.potassium - currentNPK.potassium) * adjustment),
  };

  // Scale to field size
  return {
    nitrogen: Math.round(netRequired.nitrogen * fieldSize),
    phosphorus: Math.round(netRequired.phosphorus * fieldSize),
    potassium: Math.round(netRequired.potassium * fieldSize),
  };
};

// Static method to match products
FertilizerCalculationSchema.statics.matchProducts = async function(
  requiredNPK: INPKValues,
  userId: mongoose.Types.ObjectId
): Promise<IProductRecommendation[]> {
  // This will be implemented to fetch products from database
  // For now, return a basic structure
  
  const Product = mongoose.model('Product');
  
  // Find fertilizer products
  const fertilizers = await Product.find({
    category: { $regex: /fertilizer/i },
    status: 'active',
  }).limit(20);

  const recommendations: IProductRecommendation[] = [];

  // Simple product matching logic
  // In production, this should be more sophisticated
  
  for (const product of fertilizers) {
    if (product.npk) {
      const npkParts = product.npk.split('-').map((n: string) => parseInt(n) || 0);
      if (npkParts.length === 3) {
        const [n, p, k] = npkParts;
        
        // Calculate quantity needed based on NPK content
        if (n > 0 && requiredNPK.nitrogen > 0) {
          const quantity = Math.ceil(requiredNPK.nitrogen / (n / 100));
          recommendations.push({
            productId: product._id,
            productName: product.name,
            npkRatio: product.npk,
            quantity: quantity,
            unit: 'kg',
            applicationMethod: product.applicationMethod || 'Broadcast',
            timing: 'As per schedule',
            cost: product.price ? product.price * quantity : undefined,
          });
        }
      }
    }
  }

  return recommendations.slice(0, 5); // Return top 5 recommendations
};

// Export the model
export const FertilizerCalculation: Model<IFertilizerCalculation> = 
  mongoose.models.FertilizerCalculation || 
  mongoose.model<IFertilizerCalculation>('FertilizerCalculation', FertilizerCalculationSchema);

export default FertilizerCalculation;

